### DISCUSSION TITLE
Introspection

### DISCUSSION THREAD

**Discussion by Perlkonig** [statelyai/xstate#1358] [General] [answered]
Is there a way to extract from the interpreter a list of valid events possible given the current state?

**Comment by davidkpiano** [comment, marked as answer]
If you have the current state, the next valid events are on `state.nextEvents`, which will give you an array of all "accepted" event types for that current state.

It doesn't take into account guarded transitions, since that might be dependent on event payload, which can't be known ahead of time (XState is not capable of knowing the future... yet).

**Reply by Perlkonig** [reply]
I apologize. I see that *is* in the docs. Thanks for your time.

**Comment by mashpie** [comment]
Hey, I stumbled about the same requirement and found a solution by conditionally setting event targets to `false` if they should not be allowed for a user. This way, `state.nextEvents` only reports available events. Still, one should combine this with guards, I guess. Simplified example:

```js
/**
 * kind of acl check
 */
const userAllowedTo = (capability) => true || false

/**
 * validation of message
 */
const messageValid = ({ body }) => true || false

/**
 * basic message states
 */
const stateMachine = createMachine({
  id: 'stateOfMessage',
  initial: 'created',
  states: {
    created: {
      on: {
        send: {
          target: userAllowedTo('send.Message') ? 'sent' : false,
          cond: messageValid
        },
        save: 'saved'
      }
    },
    sent: {
      delete: userAllowedTo('delete.Message') ? 'deleted' : false
    },
    save: {
      delete: 'deleted'
    }
  }
})
```

I can't tell whether this should be considered a proper solution or lead to an antipattern. Any thoughts are welcome :)
